src/Compilers/CSharp/Test/Emit2/Semantics/CollectionLiteralTests.cs

```diff
@@ -8425,6 +8655,429 @@ static void F(MyCollection<int> c)
+                    IL_00b4:  initobj    "System.Runtime.CompilerServices.TaskAwaiter<int>"
+                    IL_00ba:  ldarg.0
+                    IL_00bb:  ldc.i4.m1
+                    IL_00bc:  dup
+                    IL_00bd:  stloc.0
+                    IL_00be:  stfld      "int Program.<CreateCollection>d__1.<>1__state"
+                    IL_00c3:  ldloca.s   V_3
+                    IL_00c5:  call       "int System.Runtime.CompilerServices.TaskAwaiter<int>.GetResult()"
+                    IL_00ca:  stloc.2
+                    IL_00cb:  ldc.i4.3
+                    IL_00cc:  newarr     "int"
+                    IL_00d1:  dup
+                    IL_00d2:  ldc.i4.0
+                    IL_00d3:  ldarg.0
+                    IL_00d4:  ldfld      "int Program.<CreateCollection>d__1.<>7__wrap1"
+                    IL_00d9:  stelem.i4
+                    IL_00da:  dup
+                    IL_00db:  ldc.i4.1
+                    IL_00dc:  ldc.i4.2
+                    IL_00dd:  stelem.i4
+                    IL_00de:  dup
+                    IL_00df:  ldc.i4.2
+                    IL_00e0:  ldloc.2
+                    IL_00e1:  stelem.i4
+                    IL_00e2:  newobj     "System.ReadOnlySpan<int>..ctor(int[])"
+                    IL_00e7:  call       "MyCollection<int> MyCollectionBuilder.Create<int>(System.ReadOnlySpan<int>)"
+                    IL_00ec:  stloc.1
+                    IL_00ed:  leave.s    IL_0108
+                  }
+                  catch System.Exception
+                  {
+                    IL_00ef:  stloc.s    V_4
+                    IL_00f1:  ldarg.0
+                    IL_00f2:  ldc.i4.s   -2
+                    IL_00f4:  stfld      "int Program.<CreateCollection>d__1.<>1__state"
+                    IL_00f9:  ldarg.0
+                    IL_00fa:  ldflda     "System.Runtime.CompilerServices.AsyncTaskMethodBuilder<MyCollection<int>> Program.<CreateCollection>d__1.<>t__builder"
+                    IL_00ff:  ldloc.s    V_4
+                    IL_0101:  call       "void System.Runtime.CompilerServices.AsyncTaskMethodBuilder<MyCollection<int>>.SetException(System.Exception)"
+                    IL_0106:  leave.s    IL_011c
+                  }
+                  IL_0108:  ldarg.0
+                  IL_0109:  ldc.i4.s   -2
+                  IL_010b:  stfld      "int Program.<CreateCollection>d__1.<>1__state"
+                  IL_0110:  ldarg.0
+                  IL_0111:  ldflda     "System.Runtime.CompilerServices.AsyncTaskMethodBuilder<MyCollection<int>> Program.<CreateCollection>d__1.<>t__builder"
+                  IL_0116:  ldloc.1
+                  IL_0117:  call       "void System.Runtime.CompilerServices.AsyncTaskMethodBuilder<MyCollection<int>>.SetResult(MyCollection<int>)"
+                  IL_011c:  ret
+                }
+                """);
+        }
+
+        [ConditionalFact(typeof(CoreClrOnly))]
+        public void CollectionBuilder_AttributeCycle()
```

Comment:
I would also like to see some cycles involving `CollectionBuilderAttribute` itself. This can be in a followup PR.